### PR TITLE
MGMT-6445: Create Cluster Wizard

### PR DESCRIPTION
**Pull request by mareklibra**
Depends on:
- [x] #7  InfraEnv
- [ ] https://github.com/openshift-assisted/assisted-ui-lib/pull/624 

TODO:
- [x] YAML editor - deferred to a subtask 
- [x] postpone k8s resource creation till last step
- [x] query list of OpensShift versions
- [x] select namespace for ClusterDeployment and Secret
- [x] query list of used cluster names to perform uniqueness-validation
- [x] refactor to PF's Wizard
- [ ] Cluster Creator or Infra Owner adds a reference to the ClusterDeployment onto the InstallEnv. This reference confirms that the ClusterDeployment, which resides in a different namespace, is authorized to consume Agents from this InstallEnv. - Will be done in a follow-up

### CODE REVIEW

File: src/components/Infra/InfraEnvWizard.tsx [hunk 1]
```diff
@@ -1,164 +1,154 @@
...
+        getPullSecretResource({ namespace, name: values.name, pullSecret: values.pullSecret }),
+      );
+      await k8sCreate(
+        clusterDepModel,
+        getClusterDeployment({
+          namespace,
+          pullSecretName: secret?.metadata?.name,
+          ...values,
+        }),
+      );
```
Comment (rawagner) [inline src/components/Infra/InfraEnvWizard.tsx:48]: this needs to be 
```suggestion
        getClusterDeployment({
          namespace,
          pullSecretName: secret?.metadata?.name,
          ...values,
          labels,
        }),
      );
```
to pass the object labels from https://github.com/rawagner/dynamic-cim/pull/9/files#diff-9a7efbcdf1e2bcecc787c5040b867e71ba4851c74d2138b427d19818714e0667R35 , not the array labels from form

Comment (mareklibra) [reply src/components/Infra/InfraEnvWizard.tsx]: done

File: src/components/cluster-deployments/ClusterDeploymentWizard.tsx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    namespaced: true,
+    isList: true,
+  });
+  const usedClusterNames = (clusterDeployments || [])
+    .filter((cd) => !clusterDeployment || clusterDeployment.metadata.uid !== cd.metadata.uid)
+    .map((cd): string => `${cd.metadata.name}.${cd.spec?.baseDomain}`);
+
+  const onClusterCreate = React.useMemo(
+    () =>
+      async ({ pullSecret, openshiftVersion, ...params }: ClusterDeploymentDetailsValues) => {
```
Comment (rawagner) [inline src/components/cluster-deployments/ClusterDeploymentWizard.tsx:80]:
```suggestion
  const onClusterCreate = React.useCallback(async ({ pullSecret, openshiftVersion, ...params }: ClusterDeploymentDetailsValues) => {
```

Comment (rawagner) [reply src/components/cluster-deployments/ClusterDeploymentWizard.tsx:80]: same for other `useMemo`-s below. Not a big deal at this point though

File: src/components/cluster-deployments/ClusterDeploymentWizard.tsx [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        await onClusterDetailsUpdate(values);
+      } else {
+        await onClusterCreate(values);
+      }
+    },
+    [onClusterCreate, onClusterDetailsUpdate, clusterDeployment],
+  );
+
+  const onClose = () => {
+    history.push(`/k8s/ns/${namespace || 'all-namespaces'}/${ClusterDeploymentKind}`);
```
Comment (rawagner) [inline src/components/cluster-deployments/ClusterDeploymentWizard.tsx:183]: the `all-namespace` path is wrong. Should be just `k8s/all-namespaces` not `k8s/ns/all-namespaces`

File: src/components/cluster-deployments/EditClusterWizard.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  // it is ok if missing
+  const [agentClusterInstall] = useK8sWatchResource<AgentClusterInstallK8sResource>({
+    kind: AgentClusterInstallKind,
+    name: clusterDeployment?.spec?.clusterInstallRef?.name || '___missing__',
+    namespace,
+    namespaced: true,
+    isList: false,
+  });
+
```
Comment (rawagner) [inline src/components/cluster-deployments/EditClusterWizard.tsx:32]: if there is no `clusterInstallRef` then a better would be not do any remote request. For that you can just pass undefined as param to `useK8sWatchResource`

```suggestion
  const refName = clusterDeployment?.spec?.clusterInstallRef?.name;
  const [agentClusterInstall] = useK8sWatchResource<AgentClusterInstallK8sResource>(refName ? {
    kind: AgentClusterInstallKind,
    name: refName,
    namespace,
    namespaced: true,
    isList: false,
  } : undefined);

```

File: src/components/cluster-deployments/ClusterDeploymentWizard.tsx (hunk 2 above)
Comment (mareklibra) [reply src/components/cluster-deployments/ClusterDeploymentWizard.tsx]: done

File: src/components/cluster-deployments/ClusterDeploymentWizard.tsx (hunk 3 above)
Comment (mareklibra) [reply src/components/cluster-deployments/ClusterDeploymentWizard.tsx]: thanks, done

File: src/components/cluster-deployments/EditClusterWizard.tsx (hunk 4 above)
Comment (mareklibra) [reply src/components/cluster-deployments/EditClusterWizard.tsx]: thanks, done

File: src/components/Infra/InfraEnvForm.tsx [hunk 5]
```diff
@@ -23,117 +30,90 @@ const InfraEnvWizard: React.FC<InfraEnvWizardProps> = ({ namespace }) => {
...
+        type: 'kubernetes.io/dockerconfigjson',
+      });
+
+      const labels = values.labels.reduce((acc, curr) => {
+        const label = curr.split('=');
+        acc[label[0]] = label[1];
+        return acc;
+      }, {});
+
+      const clusterDeployment = await k8sCreate(
```
Comment (jtomasek) [inline src/components/Infra/InfraEnvForm.tsx:54]: Linter complains this is never used.

**Review by jtomasek — CHANGES_REQUESTED** [review]

File: src/components/Infra/InfraEnvForm.tsx (hunk 5 above)
Comment (mareklibra) [reply src/components/Infra/InfraEnvForm.tsx:54]: rebasing ...

**Review by jtomasek — APPROVED** [review]

**Pull request merged**